### DISCUSSION TITLE
ZS_Dead // B_CheckDeadMissionNPCs

### DISCUSSION THREAD

**Discussion by Kor-Angar** [dosinabox/g2nr_unofficial_update#488] [Q&A]
есть такая функция :-D

https://github.com/dosinabox/g2nr_unofficial_update/blob/work/PrjGOTHIC/AI/Human/ZS_Human/ZS_Dead.d

**Comment by Kor-Angar** [comment]
а вместо такой проверки `else if(Hlp_GetInstanceID(self) == Hlp_GetInstanceID(Bullco))`  можно использовать `else if C_IsNpc(SLD_807_Bullco) == true`  ?

**Reply by dosinabox** [reply]
можно, и без `== true`

**Reply by Kor-Angar** [reply]
а я так пишу что бы двойных скобок не было :-)
(C_IsNpc(SLD_807_Bullco) == true) 
(C_IsNpc(SLD_807_Bullco))

**Reply by dosinabox** [reply]
а зачем вообще скобки?
C_IsNpc(SLD_807_Bullco)

**Reply by Kor-Angar** [reply]
игра прожуёт все варианты, но условие после if вроде надо ставить в скобки

- [ ] if C_IsNpc(SLD_807_Bullco)
- [ ] if (C_IsNpc(SLD_807_Bullco))
- [ ] if (C_IsNpc(SLD_807_Bullco) == true)

**Reply by dosinabox** [reply]
после if - да, а далее скобками обрамлять не нужно, если нет операторов сравнения:
```cpp
if(C_IsNpc(SLD_807_Bullco) || (SLD_Bullco_is_alive == TRUE))
```

**Reply by dosinabox** [reply]
ну, это я так делаю)

**Reply by Kor-Angar** [reply]
вообщем в прошлом году я понял что 
```D
 if (C_IsNpc(SLD_807_Bullco) == true) // Гратт не ругается
 if C_IsNpc(SLD_807_Bullco)  // Гратт ругается
 if (C_IsNpc(SLD_807_Bullco)) // Гратт не ругается
```

ввиду того что мне двойные скобки не нравятся, я стал стараться использовать конструкцию
`if (C_IsNpc(SLD_807_Bullco) == true)`

**Comment by Kor-Angar** [comment]
в целом записывать их в мёртвые по хорошему надо в момент смерти, то есть чекин вызывать в **ZS_Dead**. На этом нам прямо указывает Гоголь в романе "Мёртвые Души" :-) Ну и значения по дефаулту желательно что бы было 0, которое меняется на 1, а в твоей версии можно его менять и на  2, так как у тебя сейчас есть Rod_KilledByPlayer и SLD_Rod_is_alive.

- 0 - нпс жив  // false
- 1 - нпс умер,  // true
- 2 - нпс убит героем  // KilledByPlayer

(SLD_Bullco_is_Dead == false) // 0 это хорошо, так если и будет глюк то последствия его будут минимальные.
(SLD_Bullco_is_alive == TRUE) // 1 это плохо, так как последствия будут максимально негативные. дети будут ныть и придётся им делать патч на коленке.

ну и по большому счёту нет смысла дробить константы НПС на 3 переменные. достаточно иметь "общего" Горна, одну переменную которая и будет отвечать за то жив он или нет, так как когда он умер  значения не имеет. как то так.

**Comment by Kor-Angar** [comment]
так вот. у тебя есть прекрасная функция B_CheckDeadMissionNPCs

в ней можно чекинить на смерть ряд НПС. фактически можно просто перенести `B_NPC_IsAliveCheck` в `B_CheckDeadMissionNPCs` поменяв/видоизменив проверки.
```D
// ******************************************************************************************************
//  			B_NPC_IsDeadCheck			(fьr NPCs die eine Levelchange vollziehen)
// ******************************************************************************************************
var int SLD_Bullco_is_dead;	//Check, ob die SLD noch leben und zum Drachenjдger gemacht werden kцnnen.
var int SLD_Rod_is_dead;
var int SLD_Cipher_is_dead;
var int SLD_Gorn_is_dead;
//var int SLD_Sylvio_is_dead;
//var int GornDJG_is_dead;
//var int DJG_Angar_is_dead;
//var int Hodges_isDead;
//var int Vino_isDead;
//var int Malak_isDead;
//var int Bromor_isDead;
//var int Fernando_isDead;
//var int Engrom_isDead;
// Sekob
// Brutus
// Randolph
// 

func void B_NPC_IsDeadCheck(var C_Npc slf)
{
    if (CurrentLevel == NEWWORLD_ZEN)
    {
        if (Kapitel <= 3)
        {
            if      C_IsNpc(slf,SLD_807_Bullco)         { SLD_Bullco_is_dead   = true; }
            else if C_IsNpc(slf,SLD_804_Rod)            { SLD_Rod_is_dead      = true; }
            else if C_IsNpc(slf,SLD_803_Cipher)         { SLD_Cipher_is_dead   = true; }
            else if C_IsNpc(slf,PC_Fighter_NW_vor_DJG)  { SLD_Gorn_is_dead     = true; };
            //else if C_IsNpc(slf,SLD_806_Sylvio)            { SLD_Sylvio_is_dead == true};
    
```

**Comment by Kor-Angar** [comment]
у тебя проверка на Равена стоит в файле с драконами?

**Reply by dosinabox** [reply]
да, в B_DragonKillCounter
